### DISCUSSION TITLE
Dedicated user object which can include groups/teams and other attributes

### DISCUSSION THREAD

**Discussion by bryan-agicap** [langfuse/langfuse#1154] [Ideas]
### Describe the feature or potential improvement

-> Users can now belong to a team/group

Uses cases:
- Could then filter by teams in the dashboard (to see cost and number of generations depending on the team per example)
- Long-term goal would be to easily generate a cost/usage report per team

### Additional information

Maybe we could just add a "group_id" to the traces/generations just like "user_id" ?

**Comment by marcklingen** [comment]
Currently `user_id` is purely a string field in Langfuse. I like the idea of modeling more complex user objects in Langfuse which can contain further attributes.

Needs:
- User optionally becomes an object instead of just a string parameter on trace() in case you want to update user attributes
- Users are updatable/queryable via the API independently from trace ingestion

Note: this is unrelated to users that have access to projects via the Langfuse UI and thus is not to be confused with the `users` db table.

**Comment by mxrcooo** [comment]
To add on to this: I have a use case where each user can belong to multiple groups, so just a single `group_id` would not be sufficient.

Having the user as an object would solve this but I imagine it would be much harder to build dashboard features that work out-of-the-box with that unless you enforce a certain structure of the object. I like the idea of having user attributes that you can use for filtering though. Posthog let's you do this for example: either via person properties (where the fields are automatically detected and available for filtering) or HogQL via person.properties.$field

**Reply by marcklingen** [reply]
thanks for the additional context on this

**Comment by marcosdd** [comment]
Would be nice to be able to set aliases to users. We are using uuid on user ids and it's not a friendly information for the business and support teams to use in the dashboard.

**Reply by marcklingen** [reply]
thanks for the +1, I'd be interested to learn more about this use case as we aim to make Langfuse more helpful for less-technical teams

**Reply by helloworld** [reply]
We would also love this. Being able to toggle in the dashboard between seeing `user_ids` and something like `emails` would be really helpful for views like this:

![CleanShot 2024-09-23 at 10 31 03@2x](https://github.com/user-attachments/assets/2643b645-52f9-4eda-8f18-9f0adf2b8048)

**Reply by marcklingen** [reply]
thanks for sharing @helloworld, makes total sense

is it an option for you to switch to emails by default to make users more readable or do you prefer to keep your internal user ids?

**Reply by ghost** [reply]
User aliases would be a very welcome addition!  
We are also using UUIDs. Having emails / other label strings would make the cost table / traces table so much more readable.

**Reply by patrikdevlin** [reply]
> Having emails / other label strings would make the cost table / traces table so much more readable.

100%, the workflow isn't great otherwise. Probably will opt to use email for userId for the time being, but being able to model a user object would be huge for readability and filtering

**Comment by misterdojo777** [comment]
what would really be useful is "tenant-id" or "client-id" to segregate different clients using our SaaS. It would be great to easily pass this information to our client's dashboards (because who doesn't want more statistics & graphs ! ;). 
Looking forward seeing this project develop. Cheers, Remo

**Comment by flefevre** [comment]
Team would be also very useful for prompt management and access management.
By defining a team of user, you can add specif prompts access.
A user could belong to multiple prompt
A prompt with a specif version could belong to multiple team
A team is a set of users and a set of prompt version.

Any idea if it will be transform as feature for the next release?

**Reply by marcklingen** [reply]
Thanks for sharing. This seems to relate more to [RBAC](https://langfuse.com/docs/rbac) than to [tracing users](https://langfuse.com/docs/tracing-features/users). Could you create a separate thread for this?

**Reply by flefevre** [reply]
I have followed your advice https://github.com/orgs/langfuse/discussions/4263
